File: source/algorithm/NQ092/NQ092_improved.py
# Python equivalent of the optimized C++ code
# 骑士最短路径问题
import sys
from collections import deque

dx = [1,2,2,1,-1,-2,-2,-1]
dy = [2,1,-1,-2,-2,-1,1,2]
n, m = 0, 0
g = []
d = []
start = (0,0)
target = (0,0)
def check(x, y):
    global n, m, g, d
    return 1 <= x <= n and 1 <= y <= m and g[x-1][y-1] != '*' and d[x][y] == -1
def find_start_and_target():
    global n, m, g, start, target
    for i in range(n):
        for j in range(m):
            if g[i][j] == 'K':
                start = (i+1, j+1)
            if g[i][j] == 'H':
                target = (i+1, j+1)
def bfs():
    global n, m, start, target, d
    d = [[-1]*(m+2) for _ in range(n+2)]
    q = deque()
    q.append(start)
    d[start[0]][start[1]] = 0
    while q:
        x, y = q.popleft()
        for i in range(8):
            nx, ny = x+dx[i], y+dy[i]
            if check(nx, ny):
                d[nx][ny] = d[x][y]+1
                q.append((nx, ny))
                if (nx, ny) == target:
                    return d[nx][ny]
    return -1
def main():
    global n, m, g
    m, n = map(int, sys.stdin.readline().split())
    g = [sys.stdin.readline().strip() for _ in range(n)]
    find_start_and_target()
    print(bfs())
if __name__ == "__main__":
    main()

File: source/algorithm/NQ092/NQ092_improved.cpp
#include <cstring>
#include <iostream>
#include <queue>
using namespace std;
// 骑士的8个可能移动方向
const int dx[8] = {1, 2, 2, 1, -1, -2, -2, -1};
const int dy[8] = {2, 1, -1, -2, -2, -1, 1, 2};
char g[160][160]; // 棋盘
int n, m;         // n是行数，m是列数
int d[160][160];  // d[x][y]表示从起点到(x,y)的最短步数
struct Point {
    int x, y;
    Point(int x = 0, int y = 0) : x(x), y(y) {}
} start, target;
queue<Point> q;
bool check(int x, int y) {
    return x >= 1 && x <= n && y >= 1 && y <= m && g[x][y] != '*' && d[x][y] == -1;
}
void findStartAndTarget() {
    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= m; ++j) {
            if (g[i][j] == 'K') {
                start.x = i;
                start.y = j;
            }
            if (g[i][j] == 'H') {
                target.x = i;
                target.y = j;
            }
        }
    }
}
int bfs() {
    memset(d, -1, sizeof(d));
    q.push(start);
    d[start.x][start.y] = 0;
    while (!q.empty()) {
        Point current = q.front();
        q.pop();
        for (int i = 0; i < 8; ++i) {
            int nx = current.x + dx[i];
            int ny = current.y + dy[i];
            if (check(nx, ny)) {
                d[nx][ny] = d[current.x][current.y] + 1;
                q.push(Point(nx, ny));
                if (nx == target.x && ny == target.y) {
                    return d[nx][ny];
                }
            }
        }
    }
    return -1;
}
int main() {
    cin >> m >> n; // 注意输入顺序：m是列数，n是行数
    for (int i = 1; i <= n; ++i) {
        cin >> (g[i] + 1); // 从g[i][1]开始存储
    }
    findStartAndTarget();
    cout << bfs() << endl;
    return 0;
}